django/utils/decorators.py

```diff
@@ -156,7 +156,47 @@ def callback(response):
                         return middleware.process_response(request, response)
                 return response
 
-            return _wrapper_view
+            @wraps(view_func)
+            async def _wrapper_view_async(request, *args, **kwargs):
+                if hasattr(middleware, "process_request"):
+                    result = middleware.process_request(request)
+                    if result is not None:
+                        return result
+                if hasattr(middleware, "process_view"):
+                    result = middleware.process_view(request, view_func, args, kwargs)
+                    if result is not None:
+                        return result
+                try:
+                    response = await view_func(request, *args, **kwargs)
+                except Exception as e:
+                    if hasattr(middleware, "process_exception"):
+                        result = middleware.process_exception(request, e)
+                        if result is not None:
+                            return result
+                    raise
+                if hasattr(response, "render") and callable(response.render):
+                    if hasattr(middleware, "process_template_response"):
+                        response = middleware.process_template_response(
+                            request, response
+                        )
+                    # Defer running of process_response until after the template
+                    # has been rendered:
+                    if hasattr(middleware, "process_response"):
+
+                        def callback(response):
+                            return middleware.process_response(request, response)
+
+                        response.add_post_render_callback(callback)
+                else:
+                    if hasattr(middleware, "process_response"):
+                        return middleware.process_response(request, response)
+                return response
```

Comment:
This duplicates almost all of the code for the `_wrapper_view_sync` function - this one is just an async function and awaits the `view_func`. I have [another PR](https://github.com/LomaxOnTheRun/django/pull/7) which looks at how we could DRY this up, as well as the repetition between the `sync_async_wrapper` and the `make_middleware_decorator` functions, but it's not a super clean refactor.

I think it offers some extra positives of sharing a code path for function decorators and class based decorators, but it feels like it should probably be done as a separate PR so a full discussion can be had for it in its own space.